graviton/blackbox.py

```diff
@@ -174,406 +196,386 @@ def _partial_encode_assert(
-        on_complete: OnComplete = OnComplete.NoOpOC,
-        dryrun_accounts: List[DryRunAccountType] = [],
-    ) -> List["DryRunInspector"]:
-        # TODO: handle txn_params
-        return list(
-            map(
-                lambda args: cls.dryrun_app(
-                    algod=algod,
-                    teal=teal,
-                    args=args,
-                    abi_method_signature=abi_method_signature,
-                    omit_method_selector=omit_method_selector,
-                    validation=validation,
-                    is_app_create=is_app_create,
-                    on_complete=on_complete,
-                    dryrun_accounts=dryrun_accounts,
-                ),
-                inputs,
-            )
-        )
+    SUGGESTED_PARAMS = SUGGESTED_PARAMS
 
-    @classmethod
-    def dryrun_logicsig_on_sequence(
-        cls,
+    def __init__(
+        self,
         algod: AlgodClient,
+        mode: ExecutionMode,
         teal: str,
-        inputs: List[Sequence[PyTypes]],
         *,
         abi_method_signature: Optional[str] = None,
-        omit_method_selector: Optional[bool] = False,
+        omit_method_selector: bool = False,
         validation: bool = True,
-    ) -> List["DryRunInspector"]:
-        # TODO: handle txn_params
-        return list(
-            map(
-                lambda args: cls.dryrun_logicsig(
-                    algod=algod,
-                    teal=teal,
-                    args=args,
-                    abi_method_signature=abi_method_signature,
-                    omit_method_selector=omit_method_selector,
-                    validation=validation,
-                ),
-                inputs,
-            )
-        )
+    ):
+        self.algod: AlgodClient = algod
+        self.mode: ExecutionMode = mode
+        self.program: str = teal
+        self.abi_method_signature: Optional[str] = abi_method_signature
+        self.omit_method_selector: bool = omit_method_selector
+        self.validation: bool = validation
+
+        self.is_app: bool
+        self.abi_argument_types: Optional[List[EncodingType]]
+        self.abi_return_type: Optional[EncodingType]
+        self.method: Optional[abi.Method]
+        self.selector: Optional[bytes]
+
+        (
+            self.is_app,
+            self.abi_argument_types,
+            self.abi_return_type,
+            self.method,
+            self.selector,
+        ) = self._init_impl(self.mode, self.abi_method_signature)
 
     @classmethod
-    def execute_one_dryrun(
-        cls,
-        algod: AlgodClient,
-        teal: str,
-        args: Sequence[PyTypes],
-        mode: ExecutionMode,
-        *,
-        abi_method_signature: Optional[str] = None,
-        omit_method_selector: Optional[bool] = False,
-        validation: bool = True,
-        txn_params: dict = {},
-        accounts: List[DryRunAccountType] = [],
-        verbose: bool = False,
-    ) -> "DryRunInspector":
+    def _init_impl(
```

Comment:
I guess you can even move forward and make it a `staticmethod`? For it is purely calculating some ABI related specs?

Reply:
This observation is correct as in this case `_init_impl()` isn't requiring anything special about it's class and therefore we could get away with a `@staticmethod`. However, I tend to be biased towards `@classmethod`'s because they are more robust against renaming of the class (or at least used to be before refactoring tools got so good). But if you feel strongly, I don't mind changing as you have suggested.

Reply:
ok, that makes sense to me, as it ties to this class